File: cli/src/commands/compile.rs
use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use crate::client::CowikiClient;
use crate::error::CliError;
use crate::output;
use crate::types::{CompileRequest, CompileResponse};
use indicatif::ProgressBar;

pub async fn run(
    client: &CowikiClient,
    branch: String,
    workspace: Option<&str>,
    timeout_secs: u64,
    json: bool,
) -> Result<(), CliError> {
    let spinner = ProgressBar::new_spinner();
    spinner.set_message("Compiling sources...");
    spinner.enable_steady_tick(Duration::from_millis(120));

    let req = CompileRequest {
        branch: branch.clone(),
    };

    let compile_fut: Pin<Box<dyn Future<Output = Result<CompileResponse, CliError>>>> = if let Some(ws) = workspace {
        Box::pin(client.compile_ws(ws, req))
    } else {
        Box::pin(client.compile(req))
    };

    let result = tokio::select! {
        r = tokio::time::timeout(Duration::from_secs(timeout_secs), compile_fut) => {
            match r {
                Ok(Ok(resp)) => Some(resp),
                Ok(Err(e)) => return Err(e),
                Err(_elapsed) => None,
            }
        }
        _ = tokio::signal::ctrl_c() => {
            spinner.finish_and_clear();
            output::print_info("Compilation request sent, server may still be processing");
            return Ok(());
        }
    };

    spinner.finish_and_clear();

    match result {
        Some(resp) => {
            if json {
                let j = serde_json::to_string_pretty(&resp.pages)
                    .map_err(|e| CliError::Unexpected(format!("JSON serialize: {e}")))?;
                println!("{j}");
            } else {
                if resp.pages.is_empty() {
                    output::print_info("No sources to compile or all already compiled.");
                    if resp.skipped > 0 {
                        output::print_info(&format!(
                            "{} source(s) skipped (already compiled).",
                            resp.skipped
                        ));
                    }
                } else {
                    let mut table = comfy_table::Table::new();
                    table.set_header(vec!["SLUG", "TITLE", "SUMMARY"]);
                    for p in &resp.pages {
                        table.add_row(vec![&p.slug, &p.title, &p.summary]);
                    }
                    output::print_table(table);
                    if resp.skipped > 0 {
                        output::print_info(&format!(
                            "{} source(s) skipped (already compiled).",
                            resp.skipped
                        ));
                    }
                }
            }
        }
        None => {
            output::print_info("Compilation still in progress, check server later");
        }
    }

    Ok(())
}

File: cli/src/commands/compile.ts
import { Command } from 'commander';
import ora from 'ora';
import { loadConfig } from '../config.js';
import { CowikiClient } from '../client.js';
import { printSuccess, printError, printInfo, printJson, printTable } from '../output.js';
import { resolveUserBranch, requireWorkspace } from '../shared.js';
import type { GlobalOpts } from '../shared.js';

export function registerCompileCommand(program: Command): void {
  program
    .command('compile')
    .description('Compile sources into wiki pages')
    .option('--timeout <seconds>', 'Timeout in seconds', '120')
    .action(async (opts, cmd) => {
      const globalOpts = cmd.parent?.optsWithGlobals() as GlobalOpts;
      const config = loadConfig(globalOpts?.server)
      const ws = requireWorkspace(globalOpts.workspace);
      const client = new CowikiClient(config.baseUrl, config.apiKey);
      const branch = await resolveUserBranch(client);

      const spinner = ora({ text: 'Compiling sources...', spinner: 'dots' }).start();

      let cancelled = false;
      const sigintHandler = () => {
        cancelled = true;
        spinner.stopAndPersist({ text: 'Compilation request sent, server may still be processing' });
      };
      process.once('SIGINT', sigintHandler);

      try {
        const controller = new AbortController();
        const timeoutMs = parseInt(opts.timeout, 10) * 1000;
        const timeoutId = setTimeout(() => controller.abort(), timeoutMs);

        // Call compile
        const compilePromise = client.compile(ws, { branch });

        // Race against timeout
        const result = await Promise.race([
          compilePromise,
          new Promise<never>((_, reject) => {
            const onAbort = () => reject(new Error('TIMEOUT'));
            controller.signal.addEventListener('abort', onAbort, { once: true });
          }),
        ]);

        clearTimeout(timeoutId);
        spinner.stop();

        if (cancelled) return;

        if (globalOpts.json) {
          printJson(result.pages);
        } else if (result.pages.length === 0) {
          printInfo('No sources to compile or all already compiled.');
          if (result.skipped > 0) {
            printInfo(`${result.skipped} source(s) skipped (already compiled).`);
          }
        } else {
          printTable(
            ['SLUG', 'TITLE', 'SUMMARY'],
            result.pages.map((p) => [p.slug, p.title, p.summary]),
          );
          if (result.skipped > 0) {
            printInfo(`${result.skipped} source(s) skipped (already compiled).`);
          }
        }
      } catch (e) {
        spinner.stop();
        const msg = e instanceof Error ? e.message : String(e);
        if (msg === 'TIMEOUT') {
          printInfo('Compilation still in progress, check server later');
        } else {
          printError(msg);
          process.exit(1);
        }
      } finally {
        process.removeListener('SIGINT', sigintHandler);
      }
    });
}
